python/zfp.pyx

```diff
@@ -0,0 +1,210 @@
+
+cdef zfp_field* _init_field(np.ndarray arr):
+    shape = arr.shape
+    cdef int ndim = arr.ndim
+    cdef zfp_type ztype = dtype_to_ztype(arr.dtype)
+    cdef zfp_field* field
+    cdef void* pointer = arr.data
+
+    strides = [int(x) / arr.itemsize for x in arr.strides[:ndim]]
+
+    if ndim == 1:
+        field = zfp_field_1d(pointer, ztype, shape[0])
+        zfp_field_set_stride_1d(field, strides[0])
+    elif ndim == 2:
+        field = zfp_field_2d(pointer, ztype, shape[0], shape[1])
+        zfp_field_set_stride_2d(field, strides[0], strides[1])
+    elif ndim == 3:
+        field = zfp_field_3d(pointer, ztype, shape[0], shape[1], shape[2])
+        zfp_field_set_stride_3d(field, strides[0], strides[1], strides[2])
+    elif ndim == 4:
+        field = zfp_field_4d(pointer, ztype, shape[0], shape[1], shape[2], shape[3])
+        zfp_field_set_stride_4d(field, strides[0], strides[1], strides[2], strides[3])
+    else:
+        raise RuntimeError("Greater than 4 dimensions not supported")
+
+    return field
+
+@cython.final
+cdef class Memory:
+    cdef void* data
+    def __cinit__(self, size_t size):
+        self.data = malloc(size)
+        if self.data == NULL:
+            raise MemoryError()
+    cdef void* __enter__(self):
+        return self.data
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        free(self.data)
+
+cpdef bytes compress_numpy(np.ndarray arr, double tolerance = -1,
+                           double rate = -1, int precision = -1,
+                           write_header=True):
+    # Input validation
+    if arr is None:
+        raise TypeError("Input array cannot be None")
+    num_params_set = sum([1 for x in [tolerance, rate, precision] if x > -1])
+    if num_params_set == 0:
+        raise ValueError("Either tolerance, rate, or precision must be set")
+    elif num_params_set > 1:
+        raise ValueError("Only one of tolerance, rate, or precision can be set")
+
+    # Setup zfp structs to begin compression
+    cdef zfp_field* field = _init_field(arr)
+    stream = zfp_stream_open(NULL)
+
+    cdef zfp_type ztype = zfp_type_none;
+    cdef int ndim = arr.ndim;
+    if tolerance > -1:
+        zfp_stream_set_accuracy(stream, tolerance)
+    elif rate > -1:
+        ztype = dtype_to_ztype(arr.dtype)
+        zfp_stream_set_rate(stream, rate, ztype, ndim, 0)
+    elif precision > -1:
+        zfp_stream_set_precision(stream, precision)
+
+    # Allocate space based on the maximum size potentially required by zfp to
+    # store the compressed array
+    cdef bytes compress_str = None
+    cdef size_t maxsize = zfp_stream_maximum_size(stream, field)
+    with Memory(maxsize) as data:
+        bstream = stream_open(data, maxsize)
+        zfp_stream_set_bit_stream(stream, bstream)
+        zfp_stream_rewind(stream)
+        # write the full header so we can reconstruct the numpy array on
+        # decompression
+        if write_header:
+            zfp_write_header(stream, field, HEADER_FULL)
```

Comment:
For safety, I'd suggest checking the return value to ensure it is not zero.

Reply:
Done! It now throws a `RuntimeError` if the return is 0.